zilencer/lib/stripe.py

```diff
@@ -178,3 +179,93 @@ def do_subscribe_customer_to_plan(stripe_customer: stripe.Customer, stripe_plan_
+    if processor.realm is None:
+        realms_with_processors = RealmAuditLog.objects.values_list('realm', flat=True)
+        query = RealmAuditLog.objects.exclude(realm__in=realms_with_processors)
+    else:
+        query = RealmAuditLog.objects.filter(realm=processor.realm)
+        global_processor = RealmAuditLog.objects.get(realm=None)
+        if global_processor.state == RealmAuditLogProcessor.SKIPPED:
+            query = query.filter(id__lte=global_processor.log_row.id)
+        else:
+            query = query.filter(id__lt=global_processor.log_row.id)
+    return query.filter(id__gt=processor.log_row.id, requires_billing_update=True) \
+                .order_by('id').first()
+
+@catch_stripe_errors
+def do_set_subscription_quantity(
+        customer: Customer, timestamp: int, idempotency_key: str, quantity: int) -> None:
+    stripe_customer = get_stripe_customer(customer.stripe_customer_id)
+    stripe_subscription = extract_current_subscription(stripe_customer)
+    stripe_subscription.quantity = quantity
+    stripe_subscription.proration_date = timestamp
+    stripe_subscription.save(idempotency_key=idempotency_key)
+
+@catch_stripe_errors
+def do_adjust_subscription_quantity(
+        customer: Customer, timestamp: int, idempotency_key: str, delta: int) -> None:
+    stripe_customer = get_stripe_customer(customer.stripe_customer_id)
+    stripe_subscription = extract_current_subscription(stripe_customer)
+    stripe_subscription.quantity = stripe_subscription.quantity + delta
+    stripe_subscription.proration_date = timestamp
+    stripe_subscription.save(idempotency_key=idempotency_key)
+
+def increment_subscription_quantity(
+        customer: Customer, timestamp: int, idempotency_key: str) -> None:
+    return do_adjust_subscription_quantity(customer, timestamp, 1)
+
+def decrement_subscription_quantity(
+        customer: Customer, timestamp: int, idempotency_key: str) -> None:
+    return do_adjust_subscription_quantity(customer, timestamp, -1)
+
+def process_next_billing_log_entry(processor: RealmAuditLogProcessor) -> bool:
+    # return value is whether there was a next thing to process
+    if processor.state in [RealmAuditLogProcessor.DONE, RealmAuditLogProcessor.SKIPPED]:
+        log_row = get_next_log_entry(processor)
+        if log_row is None:
+            return False
+        processor.log_row = log_row
+        processor.state = RealmAuditLogProcessor.STARTED
+        processor.save()
+    elif processor.state == RealmAuditLogProcessor.STARTED:
+        billing_logger.warning(
+            "Reprocessing RealmAuditLog row. realm=%s, event_type=%s, log_row.id=%s"
+            % (processor.log_row.realm.string_id, processor.log_row.event_type, processor.log_row.id))
+        log_row = processor.log_row
+    elif processor.state == RealmAuditLogProcessor.UNPROCESSED:
+        log_row = processor.log_row
+    else:
+        raise AssertionError('Unknown RealmAuditLogProcessor.state: %s' % (processor.state,))
```

Comment:
Is STALLED supposed to appear in the above?

Reply:
The STALLED ones get taken care of by
```
                for processor in RealmAuditLogProcessor.objects.exclude(
                        state=RealmAuditLogProcessor.STALLED):
```
in `process_billing_updates`

Reply:
OK.  I'd add a separate `elif/AssertionError` case here for that, then.

(It sounds like there is no code to re-process ones that are stalled yet; someone has to manually do something?)